static/app/views/dashboards/orgDashboards.tsx

```diff
@@ -98,123 +100,112 @@ class OrgDashboards extends DeprecatedAsyncComponent<Props, State> {
         // appear in the query params
-        hasSavedPageFilters(data) &&
+        hasSavedPageFilters(selectedDashboard) &&
         Object.keys(location.query).filter(unsavedQueryParam =>
           queryParamFilters.has(unsavedQueryParam)
         ).length === 0
       ) {
-        browserHistory.replace({
-          ...location,
+        navigate(
+          {
+            ...location,
+            query: {
+              ...location.query,
+              project: selectedDashboard.projects,
+              environment: selectedDashboard.environment,
+              statsPeriod: selectedDashboard.period,
+              start: selectedDashboard.start,
+              end: selectedDashboard.end,
+              utc: selectedDashboard.utc,
+            },
+          },
+          {replace: true}
+        );
+      }
+    }
+  }, [dashboardId, location, navigate, selectedDashboard]);
+
+  useEffect(() => {
+    if (!organization.features.includes('dashboards-basic')) {
+      // Redirect to Dashboards v1
+      navigate(
+        normalizeUrl({
+          pathname: `/organizations/${organization.slug}/dashboards/`,
           query: {
             ...location.query,
-            project: data.projects,
-            environment: data.environment,
-            statsPeriod: data.period,
-            start: data.start,
-            end: data.end,
-            utc: data.utc,
           },
-        });
-      }
-      return;
+        }),
+        {replace: true}
+      );
     }
+  }, [location.query, navigate, organization]);
 
-    // If we don't have a selected dashboard, and one isn't going to arrive
-    // we can redirect to the first dashboard in the list.
-    const dashboardId = data.length ? data[0].id : 'default-overview';
-    browserHistory.replace(
-      normalizeUrl({
-        pathname: `/organizations/${organization.slug}/dashboard/${dashboardId}/`,
-        query: {
-          ...location.query,
-        },
-      })
+  if (isDashboardsPending || isSelectedDashboardPending) {
+    return (
+      <Layout.Page withPadding>
+        <LoadingIndicator />
+      </Layout.Page>
     );
   }
 
-  renderLoading() {
+  if (
+    (isDashboardsPending || isSelectedDashboardPending) &&
+    selectedDashboard &&
+    hasSavedPageFilters(selectedDashboard) &&
+    Object.keys(location.query).length === 0
+  ) {
+    // Block dashboard from rendering if the dashboard has filters and
+    // the URL does not contain filters yet. The filters can either match the
+    // saved filters, or can be different (i.e. sharing an unsaved state)
     return (
       <Layout.Page withPadding>
         <LoadingIndicator />
       </Layout.Page>
     );
   }
 
-  renderBody() {
-    const {children} = this.props;
-    const {selectedDashboard, error} = this.state;
-    let dashboard = selectedDashboard;
+  if (isDashboardsError || isSelectedDashboardError) {
+    const notFound =
+      dashboardsError?.status === 404 || selectedDashboardError?.status === 404;
```

Comment:
If you're checking for an error status you'll want `retry: false` in the useApiQuery (honestly we should probably have a default setting that doesn't retry for 404s but we don't right now)